src/safeds_runner/server/memoization_map.py

```diff
@@ -0,0 +1,176 @@
+"""Module that contains the memoization logic and stats."""
+
+import logging
+import sys
+import time
+from collections.abc import Callable
+from dataclasses import dataclass
+from typing import Any
+
+
+@dataclass(frozen=True)
+class MemoizationStats:
+    """
+    Statistics calculated for every memoization call.
+
+    Parameters
+    ----------
+    last_access
+        Absolute timestamp since the unix epoch of the last access to the memoized value in nanoseconds
+    computation_time
+        Duration the computation of the value took in nanoseconds
+    lookup_time
+        Duration the lookup of the value took in nanoseconds (key comparison + IPC)
+    memory_size
+        Amount of memory the memoized value takes up in bytes
+    """
+
+    last_access: int
+    computation_time: int
+    lookup_time: int
+    memory_size: int
+
+    def __str__(self) -> str:
+        """
+        Summarizes stats contained in this object.
+
+        Returns
+        -------
+        Summary of stats
+        """
+        return (  # pragma: no cover
+            f"Last access: {self.last_access}, computation time: {self.computation_time}, lookup time:"
+            f" {self.lookup_time}, memory size: {self.memory_size}"
+        )
+
+
+class MemoizationMap:
+    """
+    The memoization map handles memoized function calls.
+
+    This contains looking up stored values, computing new values if needed and calculating and updating statistics.
+    """
+
+    def __init__(
+        self,
+        map_values: dict[tuple[str, tuple[Any], tuple[Any]], Any],
+        map_stats: dict[tuple[str, tuple[Any], tuple[Any]], MemoizationStats],
+    ):
+        """
+        Create a new memoization map using a value store dictionary and a stats dictionary.
+
+        Parameters
+        ----------
+        map_values
+            Value store dictionary
+        map_stats
+            Stats dictionary
+        """
+        self.map_values: dict[tuple[str, tuple[Any], tuple[Any]], Any] = map_values
+        self.map_stats: dict[tuple[str, tuple[Any], tuple[Any]], MemoizationStats] = map_stats
```

Comment:
It makes more sense to store the stats _per function_ instead of _per key_. Otherwise, the access to the stats already takes as long as the access to the memoization table itself. In other words, the decision whether to memoize something would already take as long as the memoization itself.